CIP-DappConnector/CIP-dapp-connector.md

```diff
@@ -0,0 +1,226 @@
+* UserDeclined - User declined to sign the transaction
+
+
+
+## Initial API
+
+In order to initiate communication from webpages to a user's Cardano wallet, the wallet must provide the following javascript functions to the webpage. These would likely be done via code injection into the webpage. The API is split into two stages to maintain the user's privacy, as the user will have to consent to `cardano_request_read_access()` in order for the dApp to read any information pertaining to the user's wallet. 
+
+### cardano_request_read_access(): Promise\<bool>
+
+Errors: APIError
+
+This is the entrypoint to start communication with the user's wallet. The wallet should request the user's permission to connect the web page to the user's wallet, and if permission has been granted, the full API (`cardano` object) must be exposed to the webpage. The wallet can choose to maintain a whitelist to not necessarily ask the user's permission every time access is requested, but this behavior is up to the wallet and should be transparent to web pages using this API. If a wallet is already connected this function should do nothing and simply return true.
+
+### cardano_check_read_access(): Promise\<bool>
+
+Errors: APIError
+
+Returns true if the full API (`cardano` object) was injected and the dApp has access to the user's wallet, and false otherwise.
+
+
+
+
+## Full API
+
+Upon successful connection via `cardano_request_read_access()`, a javascript object named `cardano` is injected with the following methods. All read-only methods (all but the signing functionality) should not require any as the user has already consented to the dApp reading information about the wallet's state when they agreed to `cardano_request_read_access()`. The remaining methods `cardano.sign_tx()`, `cardano.sign_tx_input()`, `cardano.sign_data()` must request the user's consent in an informative way for each and every API call in order to maintain security.
+
+The API chosen here is for the minimum API necessary for dApp <-> Wallet interactions without convenience functions that don't strictly need the wallet's state to work. The API here is for now also only designed for Shelley's Mary hardfork and thus has NFT support. When Alonzo is released with Plutus support this API will have to be extended.
+
+### cardano.get_utxos(amount: cbor\<value> = undefined, paginate: Paginate = undefined): Promise\<TransactionUnspentOutput[] | undefined>
+
+Errors: `APIError`, `PaginateError`
+
+If `amount` is `undefined`, this shall return a list of all UTXOs (unspent transaction outputs) controlled by the wallet. If `amount` is not `undefined`, this request shall be limited to just the UTXOs that are required to reach the combined ADA/multiasset value target specified in `amount`, and if this cannot be attained, `undefined` shall be returned. The results can be further paginated by `paginate` if it is not `undefined`.
+
+### cardano.get_balance(): Promise\<cbor\<value>>
```

Comment:
I'd suggest `get_available_balance`, or take an extra argument for specifying which balance for there are in principle multiple concepts of balances in a wallet:

- total balance (sum of all UTxO of the wallet)
- available balance (total balance, minus those in pending transactions) 
- pending balance (available balance, plus the predictive balance coming from pending change outputs)
- reward balance (balance of the reward account)

Reply:
Would we want to apply such a filter to `get_utxos()` too? At least for the pending part. The way we implemented the ergo spec in Yoroi was to just filter the utxos returned to not include ones that were pending as part of the connector's send txs. The notion of pending isn't always verifiable though, depending on how the tx was sent and to where.

Reply:
> Would we want to apply such a filter to get_utxos() too for consistency?

Yes. Incidentally, get_XXX_balance would just be the sum of all utxos returned by get_XXX_utxos.

Reply:
Yeah. We just included both endpoints since wallets would likely have the balance already calculated in the common case, and it would be such a common use-case that it would be a lot friendlier towards dApp devs to just have an endpoint for it instead of summing over the utxos every time they need the balance. Granted, initially it wasn't planned to have multiple versions of balances.